---
title: "DGE"
author: "F. Rivetti"
date: "2023-07-21"
output: html_document
---
This file details the code used for the differential gene analysis conducted in this thesis. 
```{r setup, include=FALSE}
library(DESeq2)
library(tidyverse)
library(Seurat)
library(scran)
library(pheatmap)
library(readr)
library(tibble)
library(reshape2)
library(Matrix)
library(viridis)
library(irlba)
library(stringr)
library(biomaRt)
library(SingleCellExperiment)
library(scRNAseq)
new_annotations <- readRDS("/Users/francescorivetti/Downloads/mouse_TE_age_filtered.rds")

# fix file 
new_annotations <- dplyr::select(new_annotations, -c('substitution_proportion', 'jc_distance'))
#remove unneccesary Xsomes 
new_annotations <- dplyr::filter(new_annotations, chromosome_name == "chr1" | chromosome_name == "chr2" | chromosome_name == "chr3" | chromosome_name == "chr4" | chromosome_name == "chr5" | chromosome_name == "chr6" | chromosome_name == "chr7" | chromosome_name == "chr8" | chromosome_name == "chr9" | chromosome_name == "chr10" | chromosome_name == "chr11" | chromosome_name == "chr12" | chromosome_name == "chr13" | chromosome_name == "chr14" | chromosome_name == "chr15" | chromosome_name == "chr16" | chromosome_name == "chr17" | chromosome_name == "chr18" | chromosome_name == "chr19" | chromosome_name == "chrX" | chromosome_name == "chrY")
new_annotations <- mutate(new_annotations, chromosome_name=(sub("chr", "", new_annotations$chromosome_name)))

# rename column names to mathc SeqMonk
new_annotations <- new_annotations %>% dplyr::rename(c("Chromosome" = chromosome_name, "Start" = start_position, "End" = end_position, "Subfamily" = name,  "Family" = repFamily, "Class" = repClass, "Strand" = strand, "mya" = mya ))
new_annotations$Feature <- paste(new_annotations$Chromosome, new_annotations$Start, new_annotations$End, new_annotations$Strand, sep="_")

# remove repetitive DNA
new_annotations <- dplyr::filter(new_annotations, new_annotations$Class != "Low_complexity" & new_annotations$Class != "Simple_repeat" & 
  new_annotations$Class != "tRNA" & new_annotations$Class != "snRNA"  & new_annotations$Class != "rRNA")
new_annotations <- distinct(new_annotations, Feature, .keep_all = TRUE)
rownames(new_annotations) <- new_annotations$Feature

# load previously made RDS file
sce_original <- readRDS("/Users/francescorivetti/Desktop/cello/sam_unpro/data/SCE_full.rds")

```


```{r DGE}
extracted_table <- data.frame(counts(sce_original))
condition <- unique(sce_original$Exp_setup)
coldata <- data.frame(cell_num = colnames(sce_original), condition = sce_original$Exp_setup)
######

# create Deseq object and run DGE
dds <- DESeqDataSetFromMatrix(countData = extracted_table, colData = coldata, design = ~ condition)
dds <- DESeq(dds)
# visualize plotDispEsts(dds)
# output results
res <- results(dds, alpha = 0.05)
res_tbl <- res %>%
        data.frame() %>%
        rownames_to_column(var="gene") %>%
        as_tibble()

p_cutoff <- 0.5/28634
# Subset the significant results
sig_res <- dplyr::filter(res_tbl, pvalue < p_cutoff) %>%
        dplyr::arrange(pvalue)

########### VOlcano
res_table_thres <- res_tbl %>% 
                  mutate(threshold = pvalue < p_cutoff & abs(log2FoldChange) >= 0.58)
ggplot(res_table_thres) +
    geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold)) +
    xlab("log2 fold change") + 
    ylab("-log10 adjusted p-value") +
    theme(legend.position = "right",
          plot.title = element_text(size = rel(1.5), hjust = 0.5),
          axis.title = element_text(size = rel(1.25))) + theme_bw()  



```

```{r which genes to keep}
###### expressed in a non-sparse way
sum_not_0 <- function(list) {
  sum(list!=0)
}
summary <- data.frame(apply(extracted_table, 1, sum_not_0))
summary$a <- "a"
summary_10 <- summary[summary$apply.extracted_table..1..sum_not_0. > 15, ]
genes_to_keep <- data.frame(rownames(summary_10))
colnames(genes_to_keep)[colnames(genes_to_keep) == "rownames.summary_10."] <- "gene"

acc_sig_res <- inner_join(genes_to_keep, sig_res, by='gene')

colnames(acc_sig_res)[colnames(acc_sig_res) == "gene"] <- "Feature"
anno_acc_sig_res <- inner_join(acc_sig_res, data.frame(rowData(sce_original)), by='Feature')

# manually remove overlapping genes
# Sf3b3, Cog4, Gm4673, Lsr, Gm8539mm Fbxl13, Gm14471,Nup188, Gm28038,ENSMUSG00000121395, Tmem181b-ps


```

```{r GO}

library(topGO)
library(ALL)
data(ALL)



gene_list <- dplyr::select(anno_acc_sig_res, c("Feature", "pvalue"))
# remove transposon 
gene_list <- gene_list[2:52,]
just_gene_list <- gene_list[1]
t_gene_list <- t(gene_list)
colnames(t_gene_list) <- t_gene_list[1,]
t_gene_list <- t_gene_list[2,]
gene_list_nop <- select(anno_acc_sig_res, c("Feature"))

p_values_dge <- unlist(gene_list[2])
dges <- unlist(gene_list[1])
names(p_values_dge) <- dges

## turn into dic
gene_dic <- c()

for (i in 1:nrow(gene_list)){
  gene_dic[gene_list[i,1]] <- gene_list[i,2]
}


###################
# Gene universe
library(genefilter)
gene_uni <- sce_original[rowData(sce_original)$Class == 'Gene',]
selProbes <- genefilter(logcounts(gene_uni), filterfun(pOverA(0.20, log2(100)), function(x) (IQR(x) > 0.25)))
eset <- logcounts(gene_uni)[selProbes, ]

library(org.Mm.eg.db)
# Gene universe
uni <- data.frame('gene' =rownames(eset) )
uni_p <- inner_join(uni, res_tbl[c(1,6)], by='gene')
p_values_all <- unlist(uni_p[2])
all <- unlist(uni_p[1])
names(p_values_all) <- all

##### sorry for the dumb q, how do i do that? I tried the following but it did not work
# where uni_p[2] is a vector of p-values
# uni_p[1] is a vector of gene names
# i then out all in allGenes =all
p_values_all <- unlist(uni_p[2])
all <- unlist(uni_p[1])
names(p_values_all) <- all


annotations <- select(org.Mm.eg.db, keys = unlist(gene_list[1]), keytype = "SYMBOL", columns = "GO")

GOdata <- new(
  "topGOdata",
  ontology = "BP",  # Specify the GO ontology ("BP" for Biological Process, "MF" for Molecular Function, "CC" for Cellular Component)
  allGenes =  factor(all)  , geneSel = dges,
  nodeSize = 10,  # The minimum number of genes in a GO term
  annot = annotations, 
)
```


```{r Plots}
######################################


## ggplot of top genes
normalized_log_count <- logcounts(sce_original, 
                         normalized = TRUE)
## Order results by padj values
top20_sig_genes <- acc_sig_res %>%
        dplyr::arrange(padj) %>%
        dplyr::pull(Feature) # %>%
        # head(n=27)

top20_sig_norm <- data.frame(normalized_log_count) %>%
        rownames_to_column(var = "Feature") %>%
        dplyr::filter(Feature %in% top20_sig_genes)

gathered_top20_sig <- top20_sig_norm %>%
        gather(colnames(top20_sig_norm)[2:length(colnames(top20_sig_norm))], key = "samplename", value = "log_normalized_counts")
gathered_top20_sig$condition <-  rep(sce_original$Exp_setup, each=52)       

## scatter plot 
ggplot(gathered_top20_sig) +
        geom_point(aes(x = gene, 
                       y = normalized_counts, 
                       color = condition), 
                   position=position_jitter(w=0.1,h=0)) +
        scale_y_log10() +
        xlab("Genes") +
        ylab("log10 Normalized Counts") +
        ggtitle("Top 20 Significant DE Genes") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        theme(plot.title = element_text(hjust = 0.5)) 

ggplot(gathered_top20_sig) +
        geom_boxplot(aes(x = Feature, 
                       y = normalized_counts, 
                       color = condition)) +
        scale_y_log10() +
        xlab("Genes") +
        ylab("log10 Normalized Counts") +
        ggtitle("Top 20 Significant DE Genes") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        theme(plot.title = element_text(hjust = 0.5))  



################## Heat map
order_y <- acc_sig_res %>%
        dplyr::arrange(log2FoldChange) %>%
        dplyr::pull(Feature) 
order_y[28] <- 'L1_MB5_Chr2'
library(RColorBrewer)
library(pheatmap)
pre <- gathered_top20_sig
gathered_top20_sig  <- gathered_top20_sig[order(gathered_top20_sig$condition), ]

order <- c("cell_1","cell_2", "cell_13", "cell_14", "cell_25", "cell_26", "cell_37", "cell_38", "cell_49", "cell_50", "cell_62", "cell_73", "cell_74", "cell_85", "cell_86", "cell_3","cell_4", "cell_15", "cell_16", "cell_27", "cell_28", "cell_39", "cell_40", "cell_51", "cell_52", "cell_63", "cell_64", "cell_75", "cell_76", "cell_87", "cell_88") 
order <- substring(order, 6)

# fix axis 
gathered_top20_sig$cell_number <- substring(gathered_top20_sig$samplename, 6)

ggplot(gathered_top20_sig, aes(cell_number, Feature, fill= log_normalized_counts)) +
  xlim(order) + geom_tile(color = "black") +
  scale_fill_gradient2(low = "#075AFF",
                       mid = "#FFFFCC",
                       high = "#FF0000" , midpoint = 5) +
  coord_fixed() + theme(legend.position='top') + ylab(NULL) +xlab('log2 normalized counts  mESCs with insertion ')   + ylim(order_y)   +  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

ggsave("heatmap_DGE.tiff", units="in", width=5, height=8, dpi=300, compression = 'lzw')

#############################
cols <- c('pink','yellow')
plotHeatmap(sce_original, features=anno_acc_sig_res$Feature,
    colour_columns_by=c("Exp_setup"),center=TRUE, symmetric=TRUE, zlim=c(-5, 5)) 

```

```
Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
